### DISCUSSION TITLE
[Maintenance] Choppy main menu music

### DISCUSSION THREAD

**Discussion by chrislo27** [chrislo27/PolyrhythmMania#22] [Ideas]
I've personally noticed some user-created videos of Polyrhythm Mania where the main menu music is extremely choppy/laggy.

I don't quite understand why this is, so I would like to collect some analytics to try to resolve this. If you have this issue, please comment below with your system specifications.

**In particular, I'd like to know:**
* Your operating system version (Windows 10, etc)
* Your CPU model
* RAM (Memory)
* The type of drive your system is on (hard drive HDD or solid-state SSD)
* (Optional) Your GPU model

**To find this information on Windows 10:**
1. Open Task Manager (Ctrl+Shift+Esc).
2. If you're not already in the detailed view, click the down arrow "More details".
3. Click on the Performance tab.
4. Click on the CPU information and note the full name of the CPU. You'll find it on the right-hand side, likely starting with "Intel" or "AMD".
5. Click on Memory and note down what the maximum amount is. It should say the amount on the right-hand side, for example 8.0 GB.
5. Click on "Disk 0", and note whether it says "HDD" or "SSD" below it.
6. If it is present, click on "GPU 0" and note down the name of the GPU on the right-hand side.

Example of a part of the sidebar with the GPU, memory, and main disk:
![image](https://user-images.githubusercontent.com/6299069/131951199-50f653a8-ce06-4a61-8fe0-3cdb09ea3c17.png)

**Comment by TheEggo55** [comment]
OS- Windows 10 Home v20H2
CPU- Intel(R) Core(TM) i3-7100U CPU @ 2.40GHz
Memory- 8.0 GB
Drive- SSD, SK hynix SC308 SATA 128GB
GPU- Intel(R) HD Graphics 620

**Reply by Makavo** [reply]
OS- Windows 10 Home
CPU- Intel(R) Celeron(R) N4000 CPU @ 1.10GHz
Memory- 8.0 GB
Drive- HDD, TOSHIBA MQ04ABEF100

**Comment by chrislo27** [comment]
Thanks to the folks who have submitted specs and/or helped me test some builds. More information is always appreciated!

So far it seems related to music that is read from a file and was decoded. Flushing the written bytes did not help (`FileChannel#force`). Moving the main menu music to be read all at once did result in correct music behaviour with no stuttering. It does not seem that the actual reading of file IO took a particularly long time, in fact both the already-decoded and live-decoded `MusicSample.Buffer#populate` functions took around the same average time. The time for the `AudioIO` system to prepare the next line buffer improved with the already-decoded version by a factor of nearly *40x*, with the rate of calling improving by about 7x.

**Reply by chrislo27** [reply]
The issue has been resolved with commit 92131bc2e4003bfd8755006130b399e2530b9983. Special thanks to @TheEggo55 for helping me test various builds with debugging metrics!
